supervisor/src/config.rs

```diff
@@ -0,0 +1,232 @@
+pub mod runtime {
+    pub mod v1 {
+        include!(concat!(env!("OUT_DIR"), "/encore.runtime.v1.rs"));
+    }
+}
+
+use crate::supervisor::Process;
+use base64::Engine;
+use prost::Message;
+use runtime::v1 as runtimepb;
+use std::io::Read;
+use std::time::Duration;
+use std::{collections::HashMap, env, fs::File};
+use tokio_retry::strategy::ExponentialBackoff;
+
+// loads the binary config and the runtime config and merges them into a supervisor config
+pub fn load_supervisor_config() -> SupervisorConfig {
+    let (services, gateways) = load_hosted_processes();
+    SupervisorConfig {
+        binary_config: load_binary_config(),
+        hosted_services: services,
+        hosted_gateways: gateways,
+    }
+}
+
+// loads the binary config (defaults to /encore/supervisor.config.json). It contains
+// information on how to start the available binaries and what services/gateways they contain
+fn load_binary_config() -> BinaryConfig {
+    // Parse config path from args, defaulting to /encore/supervisor.config.json if not specified
+    let args: Vec<String> = env::args().collect();
+    let config_path = args
+        .iter()
+        .position(|arg| arg == "-c")
+        .and_then(|index| args.get(index + 1))
+        .map(|s| s.to_string())
+        .unwrap_or_else(|| "/encore/supervisor.config.json".to_string());
+
+    // Open and read the supervisor config file
+    let mut file = File::open(config_path).expect("Failed to open supervisor_config.json");
+    let mut contents = String::new();
+    file.read_to_string(&mut contents)
+        .expect("Failed to read supervisor_config.json");
+    serde_json::from_str(&contents).expect("Failed to parse SupervisorConfig JSON")
+}
+
+// attempts to read the encore runtime config either as proto or json. Extracts and returns the
+// hosted services and gateways.
+fn load_hosted_processes() -> (Vec<String>, Vec<String>) {
+    // Read and decode the runtime config bytes from the environment variable
+    let runtime_config = env::var("ENCORE_RUNTIME_CONFIG")
+        .map_err(|e| format!("Failed to read ENCORE_RUNTIME_CONFIG: {}", e))
```

Comment:
This would be better handled with `.context()` from the `anyhow` crate. And then make this function return `anyhow::Result<(Vec<String>, Vec<String>)>` and don't use `.expect()` but rather `?`